### DISCUSSION TITLE
Why don't add bound methods for update types?

### DISCUSSION THREAD

**Discussion by z44d** [eternnoir/pyTelegramBotAPI#2326] [Ideas]
Adding bound methods for update types ( Message, CallbackQuery, InlineQuery, etc.. ) will make the life more easier by fill the required parameters automatically instead of fill them manually.

- This is what we use now to reply for new incoming messages:
```python
@bot.message_handler(...)
def message_handler(m):
    return bot.send_photo(m.chat.id, photo, reply_parameters=ReplyParameters(message_id=m.message_id))
```
- it would be better and more easy if there is a bound methods in telebot
```python
return m.reply_photo(photo)
```

That's applied also on all update types, i'm just wondering how famous lib like pytba doesn't support bound methods.

**Comment by coder2020official** [comment]
Hi, 
A year or so ago we had this discussion with badiboy. I am currently 50/50 on this: adding bot as optional parameter to de_json and adding some of API shortened methods wouldn't hurt the user interface. However, it may create unnecessary functions in different classes.

Let's see if @badiboy 's opinion regarding this has changed

**Comment by Badiboy** [comment]
I still hold the position that bot interface should be as close to API as it is possible. The reason is not to hurt user with unrecognized functions.

However I have nothing against creating separate module/class that can carry such kind of functions to make life easier to ones who are already deep inside the library to use it with comfort. It can even be a class that inherit Telebot / Asynctelebot and extend it with as many functions as you can imagine...

**Reply by z44d** [reply]
yeah i was thinking about create new base for update types like 

```python
class MessageWithBoundMethods:
    def reply_photo(self: "Message", photo): # and other arguments.
        return self._bot.send_photo(self.chat.id, photo, reply_parameters=ReplyParameters(message_id=self.message_id))

class Message(MessageWithBoundMethods):
    ....
```
i think it's good idea?

**Reply by z44d** [reply]
> yeah i was thinking about create new base for update types like
> 
> ```python
> class MessageWithBoundMethods:
>     def reply_photo(self: "Message", photo): # and other arguments.
>         return self._bot.send_photo(self.chat.id, photo, reply_parameters=ReplyParameters(message_id=self.message_id))
> 
> class Message(MessageWithBoundMethods):
>     ....
> ```
> 
> i think it's good idea?

and make new module contain all bases with bound methods, if you want them to stay away from types file.

**Reply by Badiboy** [reply]
> ```python
> class Message(MessageWithBoundMethods):
> ```

This way is useless, because users of Message will see all inherited from MessageWithBoundMethods. It's absolutely the same as adding methods directly to Message.

**Reply by z44d** [reply]
> > ```python
> > class Message(MessageWithBoundMethods):
> > ```
> 
> This way is useless, because users of Message will see all inherited from MessageWithBoundMethods. It's absolutely the same as adding methods directly to Message.

so add same methods as `bot.reply_to` but for all types.. like `bot.reply_phot(m, photo)` etc.. or ( `Message.methods.reply_photo` )

**Comment by z44d** [comment]
@Badiboy ping :/

**Reply by Badiboy** [reply]
> @Badiboy ping :/

Do not do like this again if you don't want me to live this thread.
